### PR TITLE
JSON input for MPM - Fix #75

### LINKED ISSUE #75: Use JSON to handle MPM input parameters
* Read input for MPM (mesh files and input parameters) from a JSON file

### PR DESCRIPTION
**Pull request by ezrayst**
First push is up, but having trouble with the initialization with `json` [input](https://github.com/ezrayst/mpm/blob/67ded92806fad1c6b8af47dedf4992d78be4eff3/tests/io_test.cc#L45) - thus `mpmtest` does not run and it will fail circleci

@kks32, can you please comment on the structure too?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -25,7 +25,7 @@
 ## Compile and Run
 > See https://mpm-doc.cb-geo.com/ for more detailed instructions. 
 
-0. Run `mkdir build && cd build && cmake -DCMAKE_BUILD_TYPE=Release /path/to/CMakeLists.txt`.
+0. Run `mkdir build && cmake -B /path/to/build/folder -DCMAKE_BUILD_TYPE=Release /path/to/CMakeLists.txt`.
```
Comment (kks32) [inline README.md]: Could you leave it as such. You create a build directory, you can `cd` to build to compile and build the files.

File: include/io.h [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+using json = nlohmann::json;
+
+//! \brief Input/Output handler
+class IO {
+ public:
+  //! Constructor
+  IO(json json_filename);
+
+  //! Return input mesh file name
+  std::string inputMeshFileName() const { return inputMeshFileName_; }
```
Comment (kks32) [inline include/io.h]: Please stick to our convention of naming variables and function names.

File: include/io.h [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+  //! Return input sub mesh file name
+  std::string inputSubMeshFileName() const { return inputSubMeshFileName_; }
```
Comment (kks32) [inline include/io.h]: Can we store the file names in a map and return the value for each key instead of multiple file names.

File: include/io.h [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+#define MPM_IO_H_
+
+#include <fstream>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <utility>
+
+#include <boost/filesystem.hpp>
+#include <eigen3/Eigen/Dense>
```
Comment (kks32) [inline include/io.h]: Why do we need eigen?

File: include/io.h [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+//! Alias for JSON
+#include "json.hpp"
+using json = nlohmann::json;
```
Comment (kks32) [inline include/io.h]: `Json`

File: include/io.h [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+  double dampingRatio() const { return dampingRatio_; }
+
+ private:
+  //! Input directory
+  std::string working_dir_;
+
+  //! Input json object
+  json json_;
+
+  //! Material properties json object
```
Comment (kks32) [inline include/io.h]: These variables are redundant if you store the `Json` object.

File: include/io.tcc [hunk 7]
```diff
@@ -0,0 +1,261 @@
+#include "io.h"
```
Comment (kks32) [inline include/io.tcc]: This is not a templatised class, so please use `.cc`

**Review by kks32 — CHANGES_REQUESTED** [review]
Please have a look at: https://github.com/cb-geo/lem/blob/develop/include/io.h

**Comment by kks32** [comment]
I would suggest just pushing your input `json` file to start with, let's agree on the input JSON, before writing rest of the code.

File: README.md (hunk 1 above)
Comment (ezrayst) [reply README.md]: OK! Understood!

**Comment by ezrayst** [comment]
@kks32, I have added JSON file as an example. Will work from this.

File: bin/mpm.json [hunk 8]
```diff
@@ -0,0 +1,30 @@
+{
+  "title": "Example JSON Input for MPM",
+  "input_directory" : "/inputFiles/",
```
Comment (kks32) [inline bin/mpm.json]: `/inputFiles/` to `./path/to/input_files/`

File: bin/mpm.json [hunk 9]
```diff
@@ -0,0 +1,30 @@
+{
+  "title": "Example JSON Input for MPM",
+  "input_directory" : "/inputFiles/", 
+  "output_directory" : "/results/",
```
Comment (kks32) [inline bin/mpm.json]: Use a `post_processing` object. Include `output_directory` and `output_steps`. Change `/results` to `./results`

File: bin/mpm.json [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  "input_files" : {
```
Comment (kks32) [inline bin/mpm.json]: Please keep all file names lower case

File: bin/mpm.json [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+    "soil_particle_filename" : "soilParticles.dat",
+    "init_stress_soil_particle_filename" : "initStressSoilP.dat",
+    "material_filename" : "material.dat",
+    "traction_filename" : "Traction.dat"
+  },
+  "analysis" : {
+    "dt" : 0.001,
+    "number_steps" : 1000,
+    "number_substeps" : 10,
+    "gravity_flag" : 0,
```
Comment (kks32) [inline bin/mpm.json]: Flags should be boolean

**Review by kks32 — CHANGES_REQUESTED** [review]

File: bin/mpm.json (hunk 11 above)
Comment (ezrayst) [reply bin/mpm.json]: Yes, thus we have 0 or 1? Or do you want "true" or "false"?

File: include/io.h [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+//! Alias for JSON
+#include "json.hpp"
+using json = nlohmann::json;
+
+//! \brief Input/Output handler
+class IO {
+ public:
+  //! Constructor
+  IO(json json_filename);
```
Comment (kks32) [inline include/io.h]: IO constructor should be `  IO(const int argc, char** argv);`

**Review by kks32 — CHANGES_REQUESTED** [review]

**Pull request merged**